Add MOBIL controller to AutomotiveSimulator

jadecastro opened the pull request:
The simulator with MOBIL planner can be run using the command (for instance):
```
bazel run //drake/automotive:demo -- -num_dragway_lanes=3 -dragway_length=500 -num_mobil_car=1 -num_trajectory_car=3
```
Currently, the implementation only supports single-segment roads (like Dragway), however it should work with any combination of `num_foo_car` and `num_dragway_lanes`.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5871)
<!-- Reviewable:end -->

jadecastro commented:
+@liangfok for feature review, please.

liangfok commented:
Checkpoint.

---

Reviewed 16 of 18 files at r1.
Review status: 16 of 18 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/automotive/automotive_demo.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8OgoXzK-sAkAE7pdI:-Ki8OgoXzK-sAkAE7pdJ:bb64e1a) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L25)):*
> ```cpp
>               "would spawn 3 cars subscribed to DRIVING_COMMAND_Russ, "
>               "DRIVING_COMMAND_Jeremy, and DRIVING_COMMAND_Liang)");
> DEFINE_int32(num_mobil_car, 0, "Number of MOBIL-controlled SimpleCar vehicles");
> ```

Missing period.

---

*[drake/automotive/automotive_demo.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8Pn6hanMzYNi5MmaB:-Ki8Pn6hanMzYNi5MmaC:b-w62l6l) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L25)):*
> ```cpp
>               "would spawn 3 cars subscribed to DRIVING_COMMAND_Russ, "
>               "DRIVING_COMMAND_Jeremy, and DRIVING_COMMAND_Liang)");
> DEFINE_int32(num_mobil_car, 0, "Number of MOBIL-controlled SimpleCar vehicles");
> ```

Should mention that this flag only applies when the terrain is a dragway.

---

*[drake/automotive/automotive_demo.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8Qqwb5ZAY9TL6NduX:-Ki8Qqwb5ZAY9TL6NduY:b-bu62qx) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L25)):*
> ```cpp
>               "would spawn 3 cars subscribed to DRIVING_COMMAND_Russ, "
>               "DRIVING_COMMAND_Jeremy, and DRIVING_COMMAND_Liang)");
> DEFINE_int32(num_mobil_car, 0, "Number of MOBIL-controlled SimpleCar vehicles");
> ```

Should mention that the MOBIL-controlled vehicles are placed behind both the idm-controlled railcars and the fixed-speed railcars.

---

*[drake/automotive/automotive_demo.cc, line 225 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8QUBeap-gRSGpm4yr:-Ki8QUBeap-gRSGpm4ys:b-k0tbdj) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L225)):*
> ```cpp
>       const double y_offset = lane->ToGeoPosition({0., 0., 0.}).y;
>       state.set_x(x_offset);
>       state.set_y(y_offset);
> ```

Is this placing all MOBIL-controlled cars at the same y-offset? If so, won't the vehicles on the same row be on top of each other?

---

*[drake/automotive/automotive_simulator.cc, line 140 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8RMAPTGo1fLyf00H0:-Ki8RMAPTGo1fLyf00H1:b-b3vexg) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.cc#L140)):*
> ```cpp
>       *road_, initial_with_s);
>   auto idm_controller =
>       builder_->template AddSystem<IdmController<T>>(*road_);
> ```

Should set the names of the above two systems.

---

*[drake/automotive/automotive_simulator.cc, line 146 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8RcvHmP2aCPF75qi7:-Ki8RcvHmP2aCPF75qi8:bdju2vu) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.cc#L146)):*
> ```cpp
>   auto pursuit = builder_->template AddSystem<PurePursuitController<T>>(*road_);
>   auto mux = builder_
>       ->template AddSystem<systems::Multiplexer<T>>(DrivingCommand<T>());
> ```

Should set the names of the above three systems.

---

*[drake/automotive/automotive_simulator.cc, line 184 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8S9yxGzyi6GRfIWUG:-Ki8S9yxGzyi6GRfIWUH:b7qm7nm) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.cc#L184)):*
> ```cpp
>   const auto& descriptor = aggregator_->AddSingleInput(name, id);
>   builder_->Connect(simple_car->pose_output(),
>                     aggregator_->get_input_port(descriptor.get_index()));
> ```

Should also wire up the SimpleCar's velocity output to the PoseAggregator. See #5894.

---

*[drake/automotive/automotive_simulator.h, line 79 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8Ssih4TlJX0Vu8rm2:-Ki8Ssih4TlJX0Vu8rm3:buosbvd) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.h#L79)):*
> ```objc
>                         const std::string& channel_name,
>                         const SimpleCarState<T>& initial_state =
>                         SimpleCarState<T>());
> ```

This undos some clang-format changes applied in #5894.

---

*[drake/automotive/automotive_simulator.h, line 86 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8T1y1Zb0sPJn-8eqJ:-Ki8T1y1Zb0sPJn-8eqK:ba73svl) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.h#L86)):*
> ```objc
>   /// @pre Start() has NOT been called.
>   ///
>   /// @param name The car's name, which must be unique among all cars. Otherwise
> ```

There is no parameter named "name".

---

*[drake/automotive/pure_pursuit_controller.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8UwV4HAuVNjjVzObK:-Ki8UwV4HAuVNjjVzObL:bdi39pn) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/pure_pursuit_controller.h#L30)):*
> ```objc
> ///   of travel.
> ///   (InputPortDescriptor getter: lane_input())
> /// Input Port 1: PoseVector for the ego car.
> ```

Doxgen does not render this correctly.

---

*[drake/automotive/pure_pursuit_controller.h, line 33 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8V3tegPpVkJ7bGgsf:-Ki8V3tegPpVkJ7bGgsg:bdeuyev) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/pure_pursuit_controller.h#L33)):*
> ```objc
> ///   (InputPortDescriptor getter: ego_pose_input())
> ///
> /// Output Port 0: A singleton BasicVector with the commanded steering angle.
> ```

BTW, I initially did not understand that "singleton BasicVector" means "a basic vector of size one". 

---

*[drake/systems/primitives/test/multiplexer_test.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8UDL8wakf6c-jngsm:-Ki8UDL8wakf6c-jngsn:bpjewe) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/systems/primitives/test/multiplexer_test.cc#L92)):*
> ```cpp
>   ASSERT_EQ(2, mux_->get_output_port(0).size());
>   ASSERT_EQ(1, output_->get_num_ports());
>   ASSERT_EQ(2, output_->get_vector_data(0)->size());
> ```

Can you also confirm the output type is a `MyVector<2, double>` and not just a `BasicVector<double>`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-:-Ki8Vv-mzkKVN6CYG14c:by96gdd)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Completed first pass. PTAL.

---

Reviewed 1 of 18 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/automotive/automotive_simulator.cc, line 139 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAU6Zd6oqHCGmx7Zq1:-KiAU6Zd6oqHCGmx7Zq2:b-qhor2e) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L139)):*
> ```cpp
>   auto mobil_planner = builder_->template AddSystem<MobilPlanner<T>>(
>       *road_, initial_with_s);
>   mobil_planner->set_name("mobil");
> ```

This name needs to be unique within the `Diagram` (include the name of the vehicle).

---

*[drake/automotive/automotive_simulator.cc, line 142 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUH71fIR-_URk2doy:-KiAUH71fIR-_URk2doz:b-qhor2e) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L142)):*
> ```cpp
>   auto idm_controller =
>       builder_->template AddSystem<IdmController<T>>(*road_);
>   idm_controller->set_name("idm");
> ```

This name needs to be unique within the `Diagram` (include the name of the vehicle).

---

*[drake/automotive/automotive_simulator.cc, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUPyHAi8Gn2B9tSHk:-KiAUPyHAi8Gn2B9tSHl:b-qhor2e) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L145)):*
> ```cpp
> 
>   auto simple_car = builder_->template AddSystem<SimpleCar<T>>();
>   simple_car->set_name(name);
> ```

This name needs to be unique within the `Diagram` (include the name of the vehicle).

---

*[drake/automotive/automotive_simulator.cc, line 149 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUInK3Bql_pvPUvMN:-KiAUInK3Bql_pvPUvMO:b-qhor2e) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L149)):*
> ```cpp
>   simple_car_initial_states_[simple_car].set_value(initial_state.get_value());
>   auto pursuit = builder_->template AddSystem<PurePursuitController<T>>(*road_);
>   pursuit->set_name("pure_pursuit");
> ```

This name needs to be unique within the `Diagram` (include the name of the vehicle).

---

*[drake/automotive/automotive_simulator.cc, line 152 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUIJNCdjLJOclMqeQ:-KiAUIJNCdjLJOclMqeR:b-qhor2e) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L152)):*
> ```cpp
>   auto mux = builder_
>       ->template AddSystem<systems::Multiplexer<T>>(DrivingCommand<T>());
>   mux->set_name("mux");
> ```

This name needs to be unique within the `Diagram` (include the name of the vehicle).

---

*[drake/automotive/test/automotive_simulator_test.cc, line 161 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAW61b1RYom08D2q_M:-KiAW61b1RYom08D2q_N:b-feii3f) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/test/automotive_simulator_test.cc#L161)):*
> ```cpp
> 
> // Returns the x-position of the vehicle based on an lcmt_viewer_draw message.
> // It also checks that the y-position of the vehicle is euqal to the provided y
> ```

typo: equal

---

*[drake/automotive/test/automotive_simulator_test.cc, line 177 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAWUe2sBAklWdzEH7W:-KiAWUe3O0ACwblSP6NF:b-7m7ubx) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/test/automotive_simulator_test.cc#L177)):*
> ```cpp
>       systems::lcm::LcmPublisherSystem::make_name(kJointStateChannelName);
> 
>   // Set up a basic simulation with just a Prius SimpleCar.
> ```

This comment should be updated to mention MOBIL.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 204 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAX1oRWtgvG1MYv3aC:-KiAX1oRWtgvG1MYv3aD:b-qwen0h) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/test/automotive_simulator_test.cc#L204)):*
> ```cpp
>   const lcmt_viewer_draw draw_message =
>       lcm->DecodeLastPublishedMessageAs<lcmt_viewer_draw>("DRAKE_VIEWER_DRAW");
>   EXPECT_LT(0., GetPosition(draw_message, 0.));
> ```

Should we place stalled vehicles some distance in front of the MOBIL-controlled SimpleCar to verify that it will switch lanes before colliding with them?

Something like this:

```
--------------------------------------------------------------
                                               | Stalled car |
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   |Ego Car|-->         | Stalled Car |
--------------------------------------------------------------
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-:-KiAY2JoOE5FYe7mKJCW:bvu844y)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 12 of 18 files reviewed at latest revision, 18 unresolved discussions.

---

*[drake/automotive/automotive_demo.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8OgoXzK-sAkAE7pdI:-Kim3bzbYKY8UsXkmWYs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L25)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Missing period.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8Pn6hanMzYNi5MmaB:-Kim3oA6Sppap_aUYkch:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L25)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should mention that this flag only applies when the terrain is a dragway.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8Qqwb5ZAY9TL6NduX:-Kim4E1-GBriPKJF8A2G:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L25)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should mention that the MOBIL-controlled vehicles are placed behind both the idm-controlled railcars and the fixed-speed railcars.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_demo.cc, line 225 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8QUBeap-gRSGpm4yr:-KimBqXIyLz0D0sp0sCH:b-ffngb6) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_demo.cc#L225)):*
Since we have:
> const Lane* lane = dragway_road_geometry->junction(0)->segment(0)->lane(lane_index);

the y-value of a unique lane lane is selected when the lane is incremented.  And when we return again to the first lane, the x-value is incremented by `kControledCarRowSpacing`.  See:
 > const double x_offset = kControlledCarRowSpacing * row;

---

*[drake/automotive/automotive_simulator.cc, line 184 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8S9yxGzyi6GRfIWUG:-KimBfNHmykNIBYEv74x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.cc#L184)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should also wire up the SimpleCar's velocity output to the PoseAggregator. See #5894.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.cc, line 139 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAU6Zd6oqHCGmx7Zq1:-KimBny_czw_sVKVV4wv:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L139)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This name needs to be unique within the `Diagram` (include the name of the vehicle).
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.cc, line 142 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUH71fIR-_URk2doy:-KimBmzMoe1tME8IJkWW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L142)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This name needs to be unique within the `Diagram` (include the name of the vehicle).
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.cc, line 145 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUPyHAi8Gn2B9tSHk:-KimBmbf2BH7XSkQxMUJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L145)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This name needs to be unique within the `Diagram` (include the name of the vehicle).
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.cc, line 149 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUInK3Bql_pvPUvMN:-KimBhaeteccoWlidN3e:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L149)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This name needs to be unique within the `Diagram` (include the name of the vehicle).
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.cc, line 152 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAUIJNCdjLJOclMqeQ:-KimBg2w-sbgETUT7cnQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/automotive_simulator.cc#L152)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This name needs to be unique within the `Diagram` (include the name of the vehicle).
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.h, line 79 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8Ssih4TlJX0Vu8rm2:-KimI6kiO9uAB6htbup6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.h#L79)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This undos some clang-format changes applied in #5894.
</blockquote></details>

Done.

---

*[drake/automotive/automotive_simulator.h, line 86 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8T1y1Zb0sPJn-8eqJ:-KimHymy9-BwKIvE9lBW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/automotive_simulator.h#L86)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

There is no parameter named "name".
</blockquote></details>

Done.

---

*[drake/automotive/pure_pursuit_controller.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8UwV4HAuVNjjVzObK:-KimGJ6ZD9Ef-01pbphe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/pure_pursuit_controller.h#L30)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Doxgen does not render this correctly.
</blockquote></details>

Done.

---

*[drake/automotive/pure_pursuit_controller.h, line 33 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8V3tegPpVkJ7bGgsf:-KimGOtvnrlaguOLgoJ0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/automotive/pure_pursuit_controller.h#L33)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, I initially did not understand that "singleton BasicVector" means "a basic vector of size one". 
</blockquote></details>

Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 161 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAW61b1RYom08D2q_M:-KimGkR4OZ78wBZxtvEW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/test/automotive_simulator_test.cc#L161)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

typo: equal
</blockquote></details>

Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 177 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAWUe2sBAklWdzEH7W:-KimHpV-H_ULGaRauZw7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/test/automotive_simulator_test.cc#L177)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This comment should be updated to mention MOBIL.
</blockquote></details>

Done.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 204 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiAX1oRWtgvG1MYv3aC:-Kime89DDbIYdULBIPJ4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/09cc7c2c8cff2cbee7476b0567255e55b17e5ed2/drake/automotive/test/automotive_simulator_test.cc#L204)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should we place stalled vehicles some distance in front of the MOBIL-controlled SimpleCar to verify that it will switch lanes before colliding with them?

Something like this:

```
--------------------------------------------------------------
                                               | Stalled car |
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   |Ego Car|-->         | Stalled Car |
--------------------------------------------------------------
```
</blockquote></details>

Done.

---

*[drake/systems/primitives/test/multiplexer_test.cc, line 92 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Ki8UDL8wakf6c-jngsm:-KimhBGso7hQOOsmLTi_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/194b7ba01d7de9aed0515a3e00fc7026f5b7f471/drake/systems/primitives/test/multiplexer_test.cc#L92)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Can you also confirm the output type is a `MyVector<2, double>` and not just a `BasicVector<double>`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 164 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiuTHjYDOhPHJd4mPI_:-KiuTHjYDOhPHJd4mPIa:b-7p4tsj) ([raw file](https://github.com/robotlocomotion/drake/blob/9c1c72d7dae18f9f64c1648104474c447fcdadc8/drake/automotive/test/automotive_simulator_test.cc#L164)):*
> ```cpp
>       systems::lcm::LcmPublisherSystem::make_name(kJointStateChannelName);
> 
>   // Set up a basic simulation with a MOBIL- and IDM-cocontrolled SimpleCar.
> ```

typo: "controlled".

---

*[drake/automotive/test/automotive_simulator_test.cc, line 183 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiuTPmLxETF05tipzb-:-KiuTPmLxETF05tipzb0:baj8a3i) ([raw file](https://github.com/robotlocomotion/drake/blob/9c1c72d7dae18f9f64c1648104474c447fcdadc8/drake/automotive/test/automotive_simulator_test.cc#L183)):*
> ```cpp
>   //  -  -  -  -  -  -  -  -  -  -  -  -  -  -    |
>   //   | MOBIL Car |   | Decoy 1 |                +---->  +s
>   // ------------------------------------------
> ```

BTW, consider marking the geo-space origin since simple car's state uses it below.

---

*[drake/systems/primitives/test/multiplexer_test.cc, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiuTyLDTUWAWbmtxbnQ:-KiuTyLDTUWAWbmtxbnR:b-xxn6hm) ([raw file](https://github.com/robotlocomotion/drake/blob/9c1c72d7dae18f9f64c1648104474c447fcdadc8/drake/systems/primitives/test/multiplexer_test.cc#L97)):*
> ```cpp
>   const auto my_vector =
>       dynamic_cast<const MyVector<2, double>*>(output_->get_vector_data(0));
>   EXPECT_NE(nullptr, my_vector);
> ```

Should use `is_dynamic_castable()` in `drake/common/test/is_dynamic_castable.h`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 15 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 164 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiuTHjYDOhPHJd4mPI_:-KivvsAJrBa5CkeK5Tzx:b-1fpyni) ([raw file](https://github.com/robotlocomotion/drake/blob/9c1c72d7dae18f9f64c1648104474c447fcdadc8/drake/automotive/test/automotive_simulator_test.cc#L164)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

typo: "controlled".
</blockquote></details>

Heh. Fixed.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 183 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiuTPmLxETF05tipzb-:-Kivwa0ALPEtmYRdLONa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9c1c72d7dae18f9f64c1648104474c447fcdadc8/drake/automotive/test/automotive_simulator_test.cc#L183)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider marking the geo-space origin since simple car's state uses it below.
</blockquote></details>

Done.

---

*[drake/systems/primitives/test/multiplexer_test.cc, line 97 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-KiuTyLDTUWAWbmtxbnQ:-Kivwb8xnGEvBlcKIG9Y:b7phi5v) ([raw file](https://github.com/robotlocomotion/drake/blob/9c1c72d7dae18f9f64c1648104474c447fcdadc8/drake/systems/primitives/test/multiplexer_test.cc#L97)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Should use `is_dynamic_castable()` in `drake/common/test/is_dynamic_castable.h`.
</blockquote></details>

Neat, thanks for pointing me to that. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-:-KiwwS_knO1BNyAFJttw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 11 of 18 files at r1, 4 of 6 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/automotive/automotive_simulator.cc, line 131 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj38X1Yx8mQ8tuYkxBb:-Kj38X1ZRLVH-sWRx97u:b-9xjr33) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/automotive_simulator.cc#L131)):*
> ```cpp
>         "AutomotiveSimulator::AddMobilControlledSimpleCar(): "
>         "RoadGeometry not set. Please call SetRoadGeometry() first before "
>         "calling this method.");
> ```

This requirement is not documented in the API.

---

*[drake/automotive/mobil_planner.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34hrgpVCwz1wviJIu:-Kj34hrgpVCwz1wviJIv:b-yz6824) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/mobil_planner.cc#L262)):*
> ```cpp
>   // TODO(jadecastro): Move this to IdmPlanner::Evaluate().
>   const T net_distance =
>       cond(delta >= T(0.), saturate(delta - idm_params.bloat_diameter(),
> ```

Please help me out by pointing out which new unit test in the PR covers this change.

---

*[drake/automotive/mobil_planner.cc, line 264 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34XknvGCnS95A8GN_:-Kj34XknvGCnS95A8GNa:bo2sxn5) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/mobil_planner.cc#L264)):*
> ```cpp
>       cond(delta >= T(0.), saturate(delta - idm_params.bloat_diameter(),
>                                    idm_params.distance_lower_limit(),
>                                    std::numeric_limits<T>::infinity()),
> ```

BTW Why use saturate with an infinite bound?  Would a bare `min` or `max` be better?

---

*[drake/automotive/mobil_planner.cc, line 265 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34MK7iBNSKyEozflJ:-Kj34MK8DcndICycMqgb:b-3sdjlo) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/mobil_planner.cc#L265)):*
> ```cpp
>                                    idm_params.distance_lower_limit(),
>                                    std::numeric_limits<T>::infinity()),
>                           saturate(delta + idm_params.bloat_diameter(),
> ```

BTW nit Vertical whitespace alignment is wrong now.

---

*[drake/automotive/pose_selector.cc, line 55 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34ryVJsBXSS-Jq1_L:-Kj34ryWDEyE1IUs53GA:b-yz6824) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/pose_selector.cc#L55)):*
> ```cpp
>           s_traffic < result_leading.pos.s) {
>         // N.B. The ego car and traffic may reside in different lanes.
>         if (s_traffic > ego_position.pos.s) {
> ```

Please help me out by pointing out which new unit test in the PR covers this change.

---

*[drake/automotive/pure_pursuit.cc, line 53 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj35SbuYyycGNV8FAnF:-Kj35SbuYyycGNV8FAnG:bgf7ele) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/pure_pursuit.cc#L53)):*
> ```cpp
> const GeoPosition PurePursuit<T>::ComputeGoalPoint(
>     const T& s_lookahead, const LaneDirection& lane_direction,
>     const RoadGeometry& road, const PoseVector<T>& pose) {
> ```

Unclear why `road` remains as a parameter.

---

*[drake/systems/primitives/multiplexer.h, line 41 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj330J-W03qCt-ww42_:-Kj330J0nf3eIQTcO5A0:b-tl070k) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/systems/primitives/multiplexer.h#L41)):*
> ```objc
>   /// Constructs a %Multiplexer with one vector-valued output port whose size
>   /// equals the size of `model_vector`.  In addition, the output type derives
>   /// from that of `model_vector`.
> ```

FYI I think the only sensible choice for input sizing is "with `model_vector.size()` scalar-valued input" so maybe that' obvious, but consider writing that out explicitly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 7 of 20 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[drake/automotive/automotive_simulator.cc, line 131 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj38X1Yx8mQ8tuYkxBb:-Kj4zct4TeMevhgviKF6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/automotive_simulator.cc#L131)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This requirement is not documented in the API.
</blockquote></details>

Done.

---

*[drake/automotive/mobil_planner.cc, line 262 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34hrgpVCwz1wviJIu:-Kj4rseuCt_g0DxcvQad:bv7v4fc) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/mobil_planner.cc#L262)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please help me out by pointing out which new unit test in the PR covers this change.
</blockquote></details>

The test case didn't exist; I have added that additional case to `mobil_planner_test.cc`. PTAL.

---

*[drake/automotive/mobil_planner.cc, line 264 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34XknvGCnS95A8GN_:-Kj4nCTz_ZbkKisWYmsK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/mobil_planner.cc#L264)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Why use saturate with an infinite bound?  Would a bare `min` or `max` be better?
</blockquote></details>

Done.

---

*[drake/automotive/mobil_planner.cc, line 265 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34MK7iBNSKyEozflJ:-Kj4pX337PZeX37G1R0u:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/mobil_planner.cc#L265)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit Vertical whitespace alignment is wrong now.
</blockquote></details>

Done.

---

*[drake/automotive/pose_selector.cc, line 55 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj34ryVJsBXSS-Jq1_L:-Kj4kqLP7S-c6UTYuQxe:bged0sf) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/pose_selector.cc#L55)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please help me out by pointing out which new unit test in the PR covers this change.
</blockquote></details>

The test case didn't exist; I have added that additional case to `pose_selector_test.cc`. PTAL.

---

*[drake/automotive/pure_pursuit.cc, line 53 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj35SbuYyycGNV8FAnF:-Kj4fHSEG1W9zbHpYl0o:bjtuwwj) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/automotive/pure_pursuit.cc#L53)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why `road` remains as a parameter.
</blockquote></details>

Removed `RoadGeometry` from both `PurePursuit` and `PurePursuitContorller`.

---

*[drake/systems/primitives/multiplexer.h, line 41 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj330J-W03qCt-ww42_:-Kj4gBsjkUI7r2mTvcTt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e71f34fed8291165e0728ffd0f9eebdd8440fb19/drake/systems/primitives/multiplexer.h#L41)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I think the only sensible choice for input sizing is "with `model_vector.size()` scalar-valued input" so maybe that' obvious, but consider writing that out explicitly.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Looks like this PR needs to be re-based with master. It contains changes that shouldn't be part of this PR.

---

Review status: 5 of 20 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-:-Kj5IUHjnLXdbtj6WPwz:bg9tt8x)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Nevermind. I was confused by Reviewable's range of diffs.

---

Review status: 5 of 20 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-:-Kj5IjAydo1mbUPP7Kxi:b-jqpj5l)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 13 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-:-Kj5QCMJSrCYOulXVFTu:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 10 of 13 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[drake/automotive/automotive_simulator.h, line 92 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj5VcqIFwWr8giel7Jk:-Kj5VcqIFwWr8giel7Jl:b-qyls5t) ([raw file](https://github.com/robotlocomotion/drake/blob/f20aebe330ee70d54dff752eb440969986a4e78a/drake/automotive/automotive_simulator.h#L92)):*
> ```objc
>   /// thrown.
>   ///
>   /// @param model_name The car's name, which must be unique among all cars.
> ```

This parameter name needs to be updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: 19 of 20 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[drake/automotive/automotive_simulator.h, line 92 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5871#-Kj5VcqIFwWr8giel7Jk:-Kj5ae_96n8wA7uJ9zwa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/f20aebe330ee70d54dff752eb440969986a4e78a/drake/automotive/automotive_simulator.h#L92)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This parameter name needs to be updated.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5871)*
<!-- Sent from Reviewable.io -->
